In [25]:
from ultralytics import YOLO


In [26]:
# Load the YOLOv11 model (using a hypothetical YOLOv11 model checkpoint "yolo11n.pt")
model = YOLO("yolo11n.pt")

In [3]:
# Train the model on your custom dataset
train_results = model.train(
    data="Cucumber fruit maturity.v3i.yolov11/data.yaml",   # Path to your custom dataset YAML file
    epochs=50,         # Number of training epochs
    imgsz=640,          # Image size for training
    device="cpu"        # Set device; use "cuda" or device="0" for GPU
)

Ultralytics 8.3.31 🚀 Python-3.10.12 torch-2.5.1+cu124 CPU (Intel Core(TM) i7-8550U 1.80GHz)
engine/trainer: task=detect, mode=train, model=yolo11n.pt, data=Cucumber fruit maturity.v3i.yolov11/data.yaml, epochs=50, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=cpu, workers=8, project=None, name=train, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True

train: Scanning /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/train/labels.cache... 1619 images, 0 backgrounds, 0 corrupt: 100%|██████████| 1619/1619 [00:00<?, ?it/s]

train: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/train/images/1191_jpg.rf.0225424c0bd8c384849b3223e82b2bb6.jpg: 1 duplicate labels removed
train: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/train/images/1403_jpg.rf.cbace095b9da30620adb380de55d3406.jpg: 1 duplicate labels removed
train: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/train/images/1444_jpg.rf.8a77d1ae3b5f971dd5db6fd1ab4787e7.jpg: 1 duplicate labels removed
train: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/train/images/1462_jpg.rf.be0dbf602b5c125c35a896bbb4a99453.jpg: 14 duplicate labels removed
train: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/train/images/1546_jpg.rf.ff185be5819fe6e15f9d31903e1b66af.jpg: 2 duplicate labels removed
train: WARNING ⚠️ /media/ola/608411B684118FA0/cv 


/media/ola/608411B684118FA0/cv project/cv_project_evn/lib/python3.10/site-packages/torch/cuda/__init__.py:129: UserWarning: CUDA initialization: CUDA unknown error - this may be due to an incorrectly set up environment, e.g. changing env variable CUDA_VISIBLE_DEVICES after program start. Setting the available devices to be zero. (Triggered internally at ../c10/cuda/CUDAFunctions.cpp:108.)
  return torch._C._cuda_getDeviceCount() > 0
val: Scanning /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/labels.cache... 204 images, 0 backgrounds, 0 corrupt: 100%|██████████| 204/204 [00:00<?, ?it/s]

val: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/images/1283_jpg.rf.dc8ecc1dfe189692dc3f9acff0c79730.jpg: 1 duplicate labels removed
val: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/images/1322_jpg.rf.829437def3f06a38fd43d3972bb3d590.jpg: 1 duplicate labels removed
val: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/images/1463_jpg.rf.e19a7bc495abc1c722b9cfb9de0ea56d.jpg: 15 duplicate labels removed
val: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/images/1570_jpg.rf.2c719d7353ee38b1f85dc3f37ac3b4c1.jpg: 3 duplicate labels removed


Plotting labels to runs/detect/train/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs/detect/train
Starting training for 50 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/50         0G      2.201      3.033      1.765         31        640: 100%|██████████| 102/102 [16:09<00:00,  9.50s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [01:15<00:00, 10.80s/it]


                   all        204       1541      0.391      0.237      0.202     0.0954

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/50         0G      1.983      2.401      1.599         32        640: 100%|██████████| 102/102 [15:24<00:00,  9.06s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:48<00:00,  6.94s/it]

                   all        204       1541      0.369      0.304       0.23      0.129



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/50         0G      1.927      2.265      1.575         37        640: 100%|██████████| 102/102 [15:08<00:00,  8.90s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:41<00:00,  5.99s/it]

                   all        204       1541      0.376      0.287      0.217      0.114



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/50         0G      1.912      2.202       1.56         29        640: 100%|██████████| 102/102 [13:46<00:00,  8.10s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:38<00:00,  5.46s/it]

                   all        204       1541      0.416       0.28      0.241      0.131



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/50         0G       1.84      2.085      1.509         41        640: 100%|██████████| 102/102 [13:43<00:00,  8.07s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:40<00:00,  5.73s/it]

                   all        204       1541      0.432      0.366       0.29       0.16



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/50         0G       1.82      2.054      1.504         47        640: 100%|██████████| 102/102 [13:42<00:00,  8.07s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:42<00:00,  6.02s/it]

                   all        204       1541      0.474      0.358       0.31      0.187



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/50         0G      1.777      1.999      1.478         32        640: 100%|██████████| 102/102 [13:32<00:00,  7.97s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:38<00:00,  5.55s/it]

                   all        204       1541      0.486      0.392      0.324      0.193



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/50         0G      1.755      1.956      1.465         35        640: 100%|██████████| 102/102 [13:31<00:00,  7.96s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:39<00:00,  5.71s/it]

                   all        204       1541      0.493      0.387      0.319      0.186



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/50         0G      1.725      1.925      1.448         39        640: 100%|██████████| 102/102 [13:32<00:00,  7.97s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:38<00:00,  5.47s/it]

                   all        204       1541       0.48      0.389      0.325      0.197



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/50         0G      1.716       1.92       1.44         38        640: 100%|██████████| 102/102 [13:29<00:00,  7.94s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:39<00:00,  5.62s/it]

                   all        204       1541       0.49      0.395      0.342       0.21



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/50         0G      1.692      1.879       1.43         14        640: 100%|██████████| 102/102 [13:30<00:00,  7.94s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:37<00:00,  5.31s/it]

                   all        204       1541      0.498      0.401      0.347      0.216



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/50         0G      1.674      1.858      1.429         51        640: 100%|██████████| 102/102 [13:24<00:00,  7.88s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:37<00:00,  5.37s/it]

                   all        204       1541      0.511      0.428      0.357      0.223



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/50         0G      1.693      1.864      1.426         52        640: 100%|██████████| 102/102 [13:27<00:00,  7.92s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:38<00:00,  5.43s/it]

                   all        204       1541      0.549      0.404      0.364      0.228



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/50         0G      1.649      1.811      1.403         36        640: 100%|██████████| 102/102 [13:23<00:00,  7.88s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:38<00:00,  5.53s/it]

                   all        204       1541      0.516      0.424      0.365      0.229



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/50         0G      1.621      1.771      1.385         31        640: 100%|██████████| 102/102 [13:24<00:00,  7.88s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:40<00:00,  5.74s/it]

                   all        204       1541      0.521      0.409      0.358      0.221



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/50         0G      1.608      1.767      1.376         27        640: 100%|██████████| 102/102 [13:21<00:00,  7.85s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:38<00:00,  5.51s/it]

                   all        204       1541      0.527      0.432      0.385      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/50         0G      1.612      1.766      1.386         39        640: 100%|██████████| 102/102 [13:25<00:00,  7.89s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:44<00:00,  6.31s/it]

                   all        204       1541      0.501      0.438      0.365      0.231



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/50         0G      1.591      1.727       1.36         42        640: 100%|██████████| 102/102 [13:36<00:00,  8.00s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:37<00:00,  5.37s/it]

                   all        204       1541      0.533      0.423       0.37      0.237



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/50         0G      1.589      1.723      1.359         38        640: 100%|██████████| 102/102 [15:39<00:00,  9.21s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:49<00:00,  7.09s/it]

                   all        204       1541      0.537       0.44      0.375      0.243



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/50         0G      1.555      1.704       1.35         30        640: 100%|██████████| 102/102 [14:00<00:00,  8.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:37<00:00,  5.40s/it]

                   all        204       1541      0.521      0.442      0.374       0.24



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/50         0G      1.565      1.709      1.361         37        640: 100%|██████████| 102/102 [12:51<00:00,  7.57s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:37<00:00,  5.34s/it]

                   all        204       1541      0.515       0.44      0.383      0.247



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/50         0G      1.573      1.719      1.364         26        640: 100%|██████████| 102/102 [13:30<00:00,  7.95s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:40<00:00,  5.85s/it]

                   all        204       1541      0.551       0.41      0.378      0.242



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/50         0G      1.539      1.675      1.339         35        640: 100%|██████████| 102/102 [14:00<00:00,  8.24s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:36<00:00,  5.23s/it]

                   all        204       1541      0.512      0.445       0.38      0.243



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/50         0G      1.534       1.65       1.32         44        640: 100%|██████████| 102/102 [12:54<00:00,  7.60s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:36<00:00,  5.28s/it]

                   all        204       1541      0.555      0.433      0.389      0.255



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/50         0G      1.542      1.674      1.342         72        640: 100%|██████████| 102/102 [12:48<00:00,  7.54s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:36<00:00,  5.22s/it]

                   all        204       1541      0.547      0.437      0.387      0.246



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/50         0G       1.53      1.657      1.324         79        640: 100%|██████████| 102/102 [12:42<00:00,  7.47s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:35<00:00,  5.11s/it]

                   all        204       1541      0.546      0.461      0.397      0.259



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/50         0G      1.519      1.634      1.315         23        640: 100%|██████████| 102/102 [12:48<00:00,  7.54s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:37<00:00,  5.42s/it]

                   all        204       1541      0.539      0.441      0.391      0.253



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/50         0G      1.506      1.613       1.31         19        640: 100%|██████████| 102/102 [12:47<00:00,  7.53s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:37<00:00,  5.32s/it]

                   all        204       1541      0.574       0.44      0.402      0.261



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/50         0G       1.52      1.633       1.32         14        640: 100%|██████████| 102/102 [12:49<00:00,  7.54s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:38<00:00,  5.45s/it]

                   all        204       1541      0.551      0.422      0.384      0.251



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/50         0G      1.477      1.576      1.292         52        640: 100%|██████████| 102/102 [12:44<00:00,  7.49s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:36<00:00,  5.21s/it]

                   all        204       1541      0.547       0.45        0.4      0.259



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      31/50         0G       1.47      1.568      1.297         28        640: 100%|██████████| 102/102 [12:41<00:00,  7.47s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:36<00:00,  5.15s/it]

                   all        204       1541      0.569      0.447      0.403      0.264



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      32/50         0G       1.47      1.561      1.297         44        640: 100%|██████████| 102/102 [12:42<00:00,  7.48s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:37<00:00,  5.29s/it]

                   all        204       1541      0.566      0.451      0.405      0.264



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      33/50         0G      1.463      1.563      1.293         44        640: 100%|██████████| 102/102 [12:47<00:00,  7.53s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:36<00:00,  5.25s/it]

                   all        204       1541      0.563      0.452      0.408      0.266



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      34/50         0G      1.479      1.572      1.295         47        640: 100%|██████████| 102/102 [11:40<00:00,  6.87s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:33<00:00,  4.75s/it]

                   all        204       1541      0.572      0.452      0.407      0.267



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      35/50         0G      1.451      1.547      1.284         51        640: 100%|██████████| 102/102 [11:49<00:00,  6.95s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:35<00:00,  5.12s/it]

                   all        204       1541      0.574      0.463      0.421      0.279



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      36/50         0G      1.437      1.508      1.274         72        640: 100%|██████████| 102/102 [11:45<00:00,  6.91s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:32<00:00,  4.62s/it]

                   all        204       1541      0.569      0.448      0.403      0.264



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      37/50         0G      1.422      1.511      1.272         21        640: 100%|██████████| 102/102 [11:48<00:00,  6.95s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:35<00:00,  5.03s/it]

                   all        204       1541      0.568      0.467       0.41      0.272



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      38/50         0G      1.429      1.512      1.272         69        640: 100%|██████████| 102/102 [11:48<00:00,  6.95s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:36<00:00,  5.24s/it]

                   all        204       1541      0.565      0.447      0.405      0.267



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      39/50         0G      1.437      1.543      1.279         17        640: 100%|██████████| 102/102 [11:43<00:00,  6.89s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:35<00:00,  5.06s/it]

                   all        204       1541      0.552      0.444      0.403      0.265



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      40/50         0G      1.417      1.493       1.26         40        640: 100%|██████████| 102/102 [11:47<00:00,  6.94s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:35<00:00,  5.06s/it]

                   all        204       1541      0.552      0.442      0.405      0.267


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      41/50         0G       1.37      1.478      1.255         20        640: 100%|██████████| 102/102 [11:35<00:00,  6.82s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:34<00:00,  4.99s/it]

                   all        204       1541      0.562      0.454      0.409      0.271



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      42/50         0G      1.349       1.44      1.257         54        640: 100%|██████████| 102/102 [11:33<00:00,  6.80s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:34<00:00,  4.90s/it]

                   all        204       1541      0.573      0.459       0.41      0.273



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      43/50         0G      1.347      1.423       1.25         24        640: 100%|██████████| 102/102 [11:30<00:00,  6.77s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:36<00:00,  5.23s/it]

                   all        204       1541       0.55       0.46      0.406      0.265



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      44/50         0G      1.326      1.388      1.242         27        640: 100%|██████████| 102/102 [11:35<00:00,  6.81s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:32<00:00,  4.57s/it]

                   all        204       1541      0.533      0.454      0.402      0.262



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      45/50         0G      1.314      1.383      1.233         19        640: 100%|██████████| 102/102 [11:34<00:00,  6.81s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:34<00:00,  4.93s/it]

                   all        204       1541      0.566      0.455      0.407      0.269



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      46/50         0G      1.317       1.36      1.226         30        640: 100%|██████████| 102/102 [11:30<00:00,  6.77s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:34<00:00,  4.99s/it]

                   all        204       1541      0.538       0.47      0.402      0.269



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      47/50         0G      1.301      1.353      1.219         15        640: 100%|██████████| 102/102 [11:32<00:00,  6.79s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:35<00:00,  5.05s/it]

                   all        204       1541       0.53       0.47      0.402      0.266



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      48/50         0G      1.302      1.339      1.227         22        640: 100%|██████████| 102/102 [11:27<00:00,  6.74s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:34<00:00,  4.89s/it]

                   all        204       1541      0.546      0.445      0.398      0.262



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      49/50         0G      1.282      1.314      1.216          8        640: 100%|██████████| 102/102 [11:31<00:00,  6.78s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:35<00:00,  5.04s/it]

                   all        204       1541      0.562      0.452      0.408      0.271



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      50/50         0G      1.276      1.295      1.205          8        640: 100%|██████████| 102/102 [11:34<00:00,  6.81s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:34<00:00,  4.86s/it]

                   all        204       1541      0.548      0.468      0.408      0.269



50 epochs completed in 11.302 hours.
Optimizer stripped from runs/detect/train/weights/last.pt, 5.5MB
Optimizer stripped from runs/detect/train/weights/best.pt, 5.5MB

Validating runs/detect/train/weights/best.pt...
Ultralytics 8.3.31 🚀 Python-3.10.12 torch-2.5.1+cu124 CPU (Intel Core(TM) i7-8550U 1.80GHz)
YOLO11n summary (fused): 238 layers, 2,582,542 parameters, 0 gradients, 6.3 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:32<00:00,  4.69s/it]


                   all        204       1541      0.575      0.462      0.422      0.279
      developing_fruit        187        667       0.59      0.493      0.449      0.297
        maturing_fruit        146        874      0.559      0.431      0.394      0.261
Speed: 1.5ms preprocess, 73.8ms inference, 0.0ms loss, 9.3ms postprocess per image
Results saved to runs/detect/train


In [ ]:
# Our trained model
model = YOLO("runs/detect/train/weights/best.pt")

In [28]:
# Evaluate model performance on the validation set
metrics = model.val()


Ultralytics 8.3.31 🚀 Python-3.10.12 torch-2.5.1+cu124 CUDA:0 (NVIDIA GeForce MX150, 4041MiB)
YOLO11n summary (fused): 238 layers, 2,582,542 parameters, 0 gradients, 6.3 GFLOPs


val: Scanning /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/labels.cache... 204 images, 0 backgrounds, 0 corrupt: 100%|██████████| 204/204 [00:00<?, ?it/s]

val: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/images/1283_jpg.rf.dc8ecc1dfe189692dc3f9acff0c79730.jpg: 1 duplicate labels removed
val: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/images/1322_jpg.rf.829437def3f06a38fd43d3972bb3d590.jpg: 1 duplicate labels removed
val: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/images/1463_jpg.rf.e19a7bc495abc1c722b9cfb9de0ea56d.jpg: 15 duplicate labels removed
val: WARNING ⚠️ /media/ola/608411B684118FA0/cv project/models/Cucumber fruit maturity.v3i.yolov11/valid/images/1570_jpg.rf.2c719d7353ee38b1f85dc3f37ac3b4c1.jpg: 3 duplicate labels removed



                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 13/13 [00:27<00:00,  2.14s/it]


                   all        204       1541      0.575      0.462      0.422      0.279
      developing_fruit        187        667       0.59      0.493      0.449      0.297
        maturing_fruit        146        874      0.559      0.431      0.394      0.261
Speed: 3.6ms preprocess, 19.1ms inference, 0.0ms loss, 31.4ms postprocess per image
Results saved to runs/detect/val2


# Summary of the Evaluation:

- The model has an average precision (mAP) of 42.2% at IoU = 0.5 across all classes, with recall of 46.2%

- Precision for developing_fruit (59%) is better than for maturing_fruit (55.9%), and the model is also performing better in terms of recall and mAP for developing fruit.




------------------

# Explaination of result:
-  cls: tensor([0.], device='cuda:0'): Indicates the class label of the detected object.

- conf: tensor([0.3029], device='cuda:0'): The model's confidence in the detection, ranging from 0 (no confidence) to 1 (maximum confidence).

- data: tensor([[4.3826e+02, 4.1274e+02, 5.0641e+02, 4.6489e+02, 3.0293e-01, 0.0000e+00]], device='cuda:0'): 

    A tensor containing the raw detection data:

        Bounding Box Coordinates (x1, y1, x2, y2):
            502.53, 218.85 (Top-left corner)
            536.29, 308.77 (Bottom-right corner)
        Confidence Score: 0.43288
        Class Label: 1.0

- id: None: This is used for tracking in scenarios where objects are tracked across multiple frames.

- is_track: False

- orig_shape: (640, 640): The original dimensions of the input image (width and height).

- shape: torch.Size([1, 6])

    The shape of the detection data tensor:
    
            1: Number of detections in this frame (1 object detected).
            6: Each detection has 6 values (x1, y1, x2, y2, confidence, class label).

- xyxy: tensor([[ 42.2078, 215.8524, 115.8837, 314.9559]], device='cuda:0'): 

        Bounding box in corner-based format:
        502.53, 218.85: Top-left corner (x1, y1)
        536.29, 308.77: Bottom-right corner (x2, y2)
